minifi/minifi-c2/minifi-c2-service/src/main/java/org/apache/nifi/minifi/c2/service/ConfigService.java

```diff
@@ -147,6 +180,140 @@ protected ConfigurationProviderInfo initContentTypeInfo(List<ConfigurationProvid
+                    .build();
+
+            Response response;
+
+            try {
+                final C2HeartbeatResponse heartbeatResponse = c2ProtocolService.processHeartbeat(heartbeat, heartbeatContext);
+                response = Response.ok(heartbeatResponse).build();
+            } catch (Exception e) {
+                logger.error("Heartbeat processing failed", e);
+                response = Response.status(BAD_REQUEST).entity(e.getMessage()).build();
+            }
+            logRequestProcessingFinished(heartbeat.getAgentId(), heartbeat.getIdentifier());
+            return response;
+        } catch (AuthorizationException e) {
+            logger.warn(HttpRequestUtil.getClientString(request) + " not authorized to access " + uriInfo, e);
+            return Response.status(403).build();
+        } catch (InvalidParameterException e) {
+            logger.info(HttpRequestUtil.getClientString(request) + " made invalid request with " + HttpRequestUtil.getQueryString(request), e);
+            return Response.status(400).entity("Invalid request.").build();
+        } catch (ConfigurationProviderException | URISyntaxException e) {
+            logger.warn("Unable to get configuration.", e);
+            return Response.status(500).build();
+        } catch (ExecutionException | UncheckedExecutionException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof WebApplicationException) {
+                throw (WebApplicationException) cause;
+            }
+            logger.error(HttpRequestUtil.getClientString(request) + " made request with " + HttpRequestUtil.getQueryString(request) + " that caused error.", cause);
+            return Response.status(500).entity("Internal error").build();
+        }
+    }
+
+    @POST
+    @Path("/acknowledge")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "An endpoint for a MiNiFi Agent to send an operation acknowledgement to the C2 server"
+    )
+    @ApiResponses({
+            @ApiResponse(code = 400, message = MESSAGE_400)})
+    public Response acknowledge(
+            @ApiParam(required = true) final C2OperationAck operationAck) {
+
+        Optional<String> agentId = getAgentId(operationAck);
+
+        agentId.ifPresent(id -> logRequestEntry(operationAck, id, operationAck.getOperationId()));
+
+        final C2ProtocolContext ackContext = C2ProtocolContext.builder()
+                .baseUri(getBaseUri())
+                .contentLength(httpServletRequest.getHeader(CONTENT_LENGTH))
+                .build();
+
+        c2ProtocolService.processOperationAck(operationAck, ackContext);
+
+        agentId.ifPresent(id -> logRequestProcessingFinished(id, operationAck.getOperationId()));
+
+        return Response.ok().build();
+
+    }
+
+    private Optional<String> getAgentId(C2OperationAck operationAck) {
+        Optional<String> agentId;
```

Comment:
You could replace the whole body of this function with
```
return Optional.ofNullable(operationAck.getAgentInfo()).map(AgentInfo::getIdentifier);
```
Those should be equivalent and it is more fluent this way in my opinion. What do you think?